File: osu!Sync/OsuReader.vb
Imports System.IO

Public Class OsuReader
    Inherits BinaryReader

    Sub New(s As Stream)
        MyBase.New(s)
    End Sub

    Public Overrides Function ReadString() As String
        Dim tag As Byte = ReadByte()
        If tag = 0 Then Return Nothing
        If tag = &HB Then Return MyBase.ReadString()
        Throw New IOException("Invalid string tag")
    End Function

    Public Function ReadDate() As Date
        Return New DateTime(ReadInt64())
    End Function
End Class

File: osu!Sync/OsuReader.cs
using System;
using System.IO;

namespace osuSync {
    public class OsuReader : BinaryReader {

		public OsuReader(Stream s) : base(s) {}

		public override string ReadString() {
			byte tag = ReadByte();
			if(tag == 0)
				return null;
			if(tag == 0xb)
				return base.ReadString();
			throw new IOException("Invalid string tag");
		}

		public DateTime ReadDate() {
			return new DateTime(ReadInt64());
		}
	}
}
